wezterm-gui/src/termwindow/mod.rs

```diff
@@ -1086,6 +1092,90 @@ impl TermWindow {
         (false, None)
     }
 
+    /// If text blinking is permitted, and it's been longer than the
+    /// normal blink rate interval, then invalidate and redraw so that
+    /// we will re-evaluate the text visibility.  This is pretty
+    /// heavyweight: it would be nice to only invalidate the lines on
+    /// which blinking text is present, and then only if some text is
+    /// within the viewport.
+    fn maintain_text_blink(&mut self, now: Instant) -> (bool, Option<Instant>) {
+        if self.focused.is_none() || self.config.text_blink_rate == 0 {
+            return (false, None);
+        }
+
+        let panes = self.get_panes_to_render();
+        if panes.is_empty() {
+            log::warn!("maintain_text_blink: get_panes_to_render.is_empty() -> close window");
+            self.window.as_ref().unwrap().close();
+            return (false, None);
+        }
+
+        for pos in panes {
+            if pos.is_active {
+                let dims = pos.pane.get_dimensions();
+                let viewport = self
+                    .get_viewport(pos.pane.pane_id())
+                    .unwrap_or(dims.physical_top);
+                let visible_range = viewport..viewport + dims.viewport_rows as StableRowIndex;
+                let (_first, lines) = pos.pane.get_lines(visible_range);
+                for line in lines {
+                    for cell in line.cells() {
+                        if cell.attrs().blink() == termwiz::cell::Blink::Slow {
+                            // A slow blinking cell is visible.
+                            let interval = Duration::from_millis(self.config.text_blink_rate);
+                            if now.duration_since(self.last_text_blink_paint) > interval {
+                                self.last_text_blink_paint = now;
+                                return (true, Some(self.last_text_blink_paint + interval));
+                            }
+                            return (false, Some(self.last_text_blink_paint + interval));
+                        }
+                    }
+                }
+            }
+        }
+
+        (false, None)
+    }
+
+    fn maintain_text_blink_rapid(&mut self, now: Instant) -> (bool, Option<Instant>) {
```

Comment:
it would be nice to combine this with the slow blink assessment so that we can avoid iterating the individual cells twice.

It's a bit unfortunate that the (presumably common case) of no blinking cells is `O(2 * viewport_area)`.   Two possible strategies to improve this:

* Add a rapid and slow blink bits to LineBits and maintain them in logically the same place as `scan_and_create_hyperlinks` and `invalidate_implicit_hyperlinks`
* Have the render logic maintain the timing similar to how animated gif frames are dispatched. https://github.com/wez/wezterm/blob/main/wezterm-gui/src/termwindow/render.rs#L80 and search for `has_animation` to get a feel for that.  [update_next_frame_time](https://github.com/wez/wezterm/blob/main/wezterm-gui/src/termwindow/render.rs#L137) is called whenever we render a cell with an anim gif.  We could do something similar when we recognize that we have a blinking cell and maintain the timing as part of that scan, avoiding additional per-cell scans

I think I like the idea of the second one more.

Reply:
It is a bit non-optimal, but it looks to me like the timer mechanism will only be checking every {interval} time, right?  So with both slow and rapid blink enabled, using the defaults of 500/250, it would be 6 scans per second (0, 250, 500, 750 for rapid) + (0, 500 for slow).

Could the has_animation method result in different timings for specific cells, such that two cells that are both rapid blink might not get updated on the same rendered frame?

Either optimization is a bit more advanced with Rust and this codebase than I am able to do right now.  Would it be worthwhile to wrap up this PR with blinking disabled by default for performance, and then optimize on a future issue/PR?